cassiopeia/solver/UPGMASolver.py

```diff
@@ -0,0 +1,213 @@
+        """Finds a pair of samples to join into a cherry.
+
+        Finds the pair of samples with the minimum dissimilarity by finding the
+        minimum value in the provided dissimilarity matrix
+
+        Args:
+            dissimilarity_matrix: A sample x sample dissimilarity matrix
+
+        Returns:
+            A tuple of integers representing rows in the dissimilarity matrix
+                to join.
+        """
+
+        dissimilarity_matrix = dissimilarity_matrix.astype(float)
+        np.fill_diagonal(dissimilarity_matrix, np.inf)
+
+        return np.unravel_index(np.argmin(dissimilarity_matrix, axis=None), dissimilarity_matrix.shape)
+
+
+    def update_dissimilarity_map(
+        self,
+        dissimilarity_map: pd.DataFrame,
+        cherry: Tuple[str, str],
+        new_node: str,
+        cluster_to_cluster_size: Dict[str, int]
+    ) -> pd.DataFrame:
+        """Update dissimilarity map after finding a cherry.
+
+        Updates the dissimilarity map after joining together two nodes (m1, m2)
+        at a cherry m. For all other nodes v, the new dissimilarity map d' is:
+
+        d'(m, v) = (|m1| * d(m1, v) + |m2| * d(m2, v))/(|m1| + |m2|)
+
+        where |m1| is the size of cluster m1, i.e. the number of sample leaves
+        under node m1.
+
+        Args:
+            dissimilarity_map: A dissimilarity map to update
+            cherry: A tuple of indices in the dissimilarity map that are joining
+            new_node: New node name, to be added to the new dissimilarity map
+
+        Returns:
+            A new dissimilarity map, updated with the new node
+        """
+        i_size, j_size = cluster_to_cluster_size[cherry[0]], cluster_to_cluster_size[cherry[1]]
+        cluster_to_cluster_size[new_node] = i_size + j_size
+
+        i, j = (
+            np.where(dissimilarity_map.index == cherry[0])[0][0],
+            np.where(dissimilarity_map.index == cherry[1])[0][0],
+        )
+
+        dissimilarity_array = self.update_dissimilarity_map_numba(
+            dissimilarity_map.to_numpy(), i, j, i_size, j_size
+        )
+        sample_names = list(dissimilarity_map.index) + [new_node]
+
+        dissimilarity_map = pd.DataFrame(
+            dissimilarity_array, index=sample_names, columns=sample_names
+        )
+
+        # drop out cherry from dissimilarity map
+        dissimilarity_map.drop(
+            columns=[cherry[0], cherry[1]],
+            index=[cherry[0], cherry[1]],
+            inplace=True,
+        )
+
+        return dissimilarity_map
+
+    @staticmethod
+    @numba.jit(nopython=True)
+    def update_dissimilarity_map_numba(
+        dissimilarity_map: np.array, cherry_i: int, cherry_j: int, size_i: int, size_j: int
+    ) -> np.array:
+        """An optimized function for updating dissimilarities.
+
+        A faster implementation of updating the dissimilarity map for Neighbor
+        Joining, invoked by `self.update_dissimilarity_map`.
```

Comment:
"A faster implementation... for **UPGMA...**"

Reply:
Whoops thanks again!